04-onion-routing.md

```diff
@@ -463,6 +463,163 @@ Having the shared secrets of all intermediate nodes it can unwrap the packet unt
+
+If a node has requirement advertised in its `node_announcement`
+`features` which were not present in the onion:
+
+1. type: PERM|NODE|3 (`required_node_feature_missing`)
+
+A forwarding node MAY return one of the following errors, the final
+node MUST NOT:
+
+If the onion version byte is unknown:
+
+1. type: BADONION|PERM|4 (`invalid_onion_version`)
+2. data:
+   * [32:sha256-of-onion]
+
+If the ephemeral key in the onion is invalid:
+
+1. type: BADONION|PERM|5 (`invalid_onion_key`)
+2. data:
+   * [32:sha256-of-onion]
+
+If an otherwise unspecified transient error occurs for the outgoing
+channel (eg. peer unresponsive):
+
+1. type: 6 (`temporary_channel_failure`)
+
+If an otherwise unspecified permanent error occurs for the outgoing
+channel (eg. channel (recently) closed):
+
+1. type: PERM|7 (`permanent_channel_failure`)
+
+If the outgoing channel has requirement advertised in its
+`channel_announcement` `features` which were not present in the onion:
+
+1. type: PERM|8 (`required_channel_feature_missing`)
+
+If the next peer specified by the onion is not known:
+
+1. type: PERM|9 (`unknown_next_peer`)
+
+If the HTLC does not reach the current minimum amount, we tell them
+the amount of the incoming HTLC and the current channel setting for
+the outgoing channel:
+
+1. type: UPDATE|10 (`amount_below_minimum`)
+2. data:
+   * [4:htlc-msat]
+   * [2:len]
+   * [len:channel_update]
+
+If the HTLC does not pay sufficient fee, we tell them the amount of
+the incoming HTLC and the current channel setting for the outgoing
+channel:
+
+1. type: UPDATE|11 (`fee_insufficient`)
+2. data:
+   * [4:htlc-msat]
+   * [2:len]
+   * [len:channel_update]
+
+If `outgoing_cltv_value` does not match the`update_add_htlc`'s
+`cltv-expiry` minus `cltv-expiry-delta` for the outgoing channel, we
+tell them the `cltv-expiry` and the current channel setting for the
+outgoing channel:
+
+1. type: UPDATE|12 (`incorrect_cltv_expiry`)
+2. data:
+   * [4:cltv-expiry]
+   * [2:len]
+   * [len:channel_update]
+
+If the ctlv-expiry is too near, we tell them the the current channel
+setting for the outgoing channel:
+
+1. type: UPDATE|13 (`expiry_too_soon`)
+2. data:
+   * [2:len]
+   * [len:channel_update]
+
+The final node may return one of the following errors, intermediate
+nodes MUST NOT:
+
+If the payment hash has already been paid, the final node MAY treat
+the payment hash as unknown, or may succeed in accepting the HTLC.
+If the payment hash is unknown, the final node MUST fail the HTLC:
+
+1. type: PERM|13 (`unknown_payment_hash`)
+
+If the amount paid is less than the amount expected, the final node
+MUST fail the HTLC.  If the amount paid is more than the amount
+expected, the final node SHOULD fail the HTLC:
+
+1. type: PERM|14 (`incorrect_payment_amount`)
+
+If the `cltv-expiry` is too low, the final node MUST fail the HTLC:
+
+1. type: 15 (`final_expiry_too_soon`)
+
+### Receiving Failure Codes
+
+If the sending node is the final node:
```

Comment:
We should specify that this is the node sending the error, not the node initiating the transfer.